### PR TITLE
Triangle element class and unit tests

### PR DESCRIPTION
**Pull request by thiagordonho**
Implement the `TriangleElement` class that includes:

- 3-noded triangular element.
- 6-noded triangular element.

And the quadrature rules for such elements (`TriangleQuadrature `class):

- 1 quadrature point.
- 3 quadrature points.

The unit tests for these classes are also implemented.

### CODE REVIEW

File: include/cell.tcc [hunk 1]
```diff
@@ -450,9 +469,17 @@ inline bool mpm::Cell<Tdim>::is_point_in_cell(
   else
     (*xi) = this->transform_real_to_unit_cell(point);
 
-  // Check if the transformed coordinate is within the unit cell (-1, 1)
-  for (unsigned i = 0; i < (*xi).size(); ++i)
-    if ((*xi)(i) < -1. || (*xi)(i) > 1. || std::isnan((*xi)(i))) status = false;
+  // Check if the transformed coordinate is within the unit cell:
+  // between 0 and 1-xi(1-i) if the element is a triangle, and between
+  // -1 and 1 if otherwise
+  switch (this->element_->corner_indices().size()) {
```
Comment (kks32) [inline include/cell.tcc]: Why can't you have an `if else`?

**Review by kks32 — CHANGES_REQUESTED** [review]

File: include/cell.tcc (hunk 1 above)
Comment (thiagordonho) [reply include/cell.tcc]: I thought about it. But if in the future someone wants to implement another type of element than this structure would be more helpful. Should I still change it to if-else?

File: include/cell.tcc [hunk 2]
```diff
@@ -223,13 +223,34 @@ inline void mpm::Cell<1>::compute_volume() {
...
+      //     |        `\
+      //   0 0----------0 1
+      //
+      auto node0 = nodes_[indices(0)]->coordinates();
+      auto node1 = nodes_[indices(1)]->coordinates();
+      auto node2 = nodes_[indices(2)]->coordinates();
+
+      // 2 * Area = (x1 * y2 - x2 * y1) - (x0 * y2 - x2 * y0) + (x0 * y1 - x1 *
+      // y0)
+      volume_ = std::fabs(((node1(0) * node2(1)) - (node2(0) * node1(1))) -
```
Comment (kks32) [inline include/cell.tcc:249]: This is confusing, if we are returning the `volume`, we should return the `area` and not `2 * area`

File: include/cell.tcc [hunk 3]
```diff
@@ -450,9 +471,18 @@ inline bool mpm::Cell<Tdim>::is_point_in_cell(
   else
     (*xi) = this->transform_real_to_unit_cell(point);
 
-  // Check if the transformed coordinate is within the unit cell (-1, 1)
-  for (unsigned i = 0; i < (*xi).size(); ++i)
-    if ((*xi)(i) < -1. || (*xi)(i) > 1. || std::isnan((*xi)(i))) status = false;
+  // Check if the transformed coordinate is within the unit cell:
+  // between 0 and 1-xi(1-i) if the element is a triangle, and between
+  // -1 and 1 if otherwise
+  if (this->element_->corner_indices().size() == 3) {
```
Comment (kks32) [inline include/cell.tcc:478]:
```
double min = -1., max = 1.;
if (this->element_->corner_indices().size() == 3) {
  min = 0;
  max = 1.;
}
for (unsigned i = 0; i < (*xi).size(); ++i)
  if ((*xi)(i) < min || (*xi)(i) > max || std::isnan((*xi)(i)))
    status = false;
```

Comment (kks32) [reply include/cell.tcc:478]: I know this doesn't test the inclined side, but we need a better approach than too many if else conditions

File: include/cell.tcc [hunk 4]
```diff
@@ -502,8 +532,35 @@ inline Eigen::Matrix<double, 2, 1> mpm::Cell<2>::local_coordinates_point(
...
+
+      auto node0 = nodes_[indices(0)]->coordinates();
+      auto node1 = nodes_[indices(1)]->coordinates();
+      auto node2 = nodes_[indices(2)]->coordinates();
+
+      const double area = ((node1(0) - node0(0)) * (node2(1) - node0(1)) -
+                           (node2(0) - node0(0)) * (node1(1) - node0(1))) /
+                          2.0;
+
+      xi(0) = 1 / (2 * area) *
```
Comment (kks32) [inline include/cell.tcc]: Use `1./ (2. * area)` when using floating points

File: include/triangle_element.tcc [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+    const Eigen::Matrix<double, 2, 1>& particle_size,
+    const Eigen::Matrix<double, 2, 1>& deformation_gradient) const {
+  Eigen::Matrix<double, 3, 1> shapefn;
+  shapefn(0) = 1 - (xi(0) + xi(1));
+  shapefn(1) = xi(0);
+  shapefn(2) = xi(1);
+  return shapefn;
+}
+
+//! Return gradient of shape functions of a 4-node Triangle Element at a
```
Comment (kks32) [inline include/triangle_element.tcc]:
```suggestion
//! Return gradient of shape functions of a 3-node Triangle Element at a
```

File: include/cell.tcc (hunk 3 above)
Comment (kks32) [reply include/cell.tcc:478]: Alternatively, we call a function in `element` class to see if a point is in it's unit cell, that would be a lot nicer. I'd go with that. If you wish that could be a separate PR improving this. So leave this section as is, and then let's do a separate PR improving this.

File: include/cell.tcc (hunk 2 above)
Comment (kks32) [reply include/cell.tcc:249]: Where are you using this?

**Review by kks32 — CHANGES_REQUESTED** [review]

**Review by kks32 — APPROVED** [review]

**Pull request merged**